src/vm/eventpipebuffermanager.cpp

```diff
@@ -0,0 +1,595 @@
+    }
+
+    // See if the thread already has a buffer to try.
+    bool allocNewBuffer = false;
+    EventPipeBuffer *pBuffer = NULL;
+    EventPipeBufferList *pThreadBufferList = pThread->GetEventPipeBufferList();
+    if(pThreadBufferList == NULL)
+    {
+        allocNewBuffer = true;
+    }
+    else
+    {
+        // The thread already has a buffer list.  Select the newest buffer and attempt to write into it.
+        pBuffer = pThreadBufferList->GetTail();
+        if(pBuffer == NULL)
+        {
+            // This should never happen.  If the buffer list exists, it must contain at least one entry.
+            _ASSERT(!"Thread buffer list with zero entries encountered.");
+            return false;
+        }
+        else
+        {
+            // Attempt to write the event to the buffer.  If this fails, we should allocate a new buffer.
+            allocNewBuffer = !pBuffer->WriteEvent(pEventThread, event, pData, length, pStack);
+        }
+    }
+
+    // Check to see if we need to allocate a new buffer, and if so, do it here.
+    if(allocNewBuffer)
+    {
+        unsigned int requestSize = sizeof(EventPipeEventInstance) + length;
+        pBuffer = AllocateBufferForThread(pThread, requestSize);
+    }
+
+    // Try to write the event after we allocated (or stole) a buffer.
+    // This is the first time if the thread had no buffers before the call to this function.
+    // This is the second time if this thread did have one or more buffers, but they were full.
+    if(allocNewBuffer && pBuffer != NULL)
+    {
+        allocNewBuffer = !pBuffer->WriteEvent(pEventThread, event, pData, length, pStack);
+    }
+
+    return !allocNewBuffer;
+}
+
+void EventPipeBufferManager::WriteAllBuffersToFile(EventPipeFile *pFile, LARGE_INTEGER stopTimeStamp)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+        PRECONDITION(pFile != NULL);
+    }
+    CONTRACTL_END;
+
+    // TODO: Better version of merge sort.
+    // 1. Iterate through all of the threads, adding each buffer to a temporary list.
+    // 2. While iterating, get the lowest most recent timestamp.  This is the timestamp that we want to process up to.
+    // 3. Process up to the lowest most recent timestamp for the set of buffers.
+    // 4. When we get NULLs from each of the buffers on PopNext(), we're done.
+    // 5. While iterating if PopNext() == NULL && Empty() == NULL, remove the buffer from the list.  It's empty.
+    // 6. While iterating, grab the next lowest most recent timestamp.
+    // 7. Walk through the list again and look for any buffers that have a lower most recent timestamp than the next most recent timestamp.
+    // 8. If we find one, add it to the list and select its most recent timestamp as the lowest.
+    // 9. Process again (go to 3).
+    // 10. Continue until there are no more buffers to process.
+
+    // Niavely walk the circular buffer, writing the event stream in timestamp order.
+    while(true)
```

Comment:
Yes, this is true.  Since this is a v1 feature and we expect event loads to be relatively low for the scenario that we're targeting, this is likely to be OK.  Going forward, I agree that we do need to look at this.

I will file an issue for this.